File: sol/2658/2658.py
class Solution:
    def findMaxFish(self, grid: List[List[int]]) -> int:
        m = len(grid)
        n = len(grid[0])
        res = 0
        for i in range(m):
            for j in range(n):
                if grid[i][j] <= 0:
                    continue
                total = 0
                buff = [(i, j)]
                while len(buff) != 0:
                    x, y = buff.pop()
                    if grid[x][y] <= 0:
                        continue
                    total += grid[x][y]
                    grid[x][y] = 0
                    for dx, dy in ((0, 1), (0, -1), (1, 0), (-1, 0)):
                        xx = x + dx
                        yy = y + dy
                        if 0 <= xx < m and 0 <= yy < n and grid[xx][yy] > 0:
                            buff.append((xx, yy))
                res = max(res, total)
        return res

File: sol/2658/2658.cpp
struct Solution {
    static constexpr int DIRS[4][2] = {
        {1, 0},
        {-1, 0},
        {0, 1},
        {0, -1}
    };
    int findMaxFish(vector<vector<int>>& grid) {
        int m = grid.size(), n = grid[0].size(), res = 0;
        for (int i = 0; i < m; ++i){
            for (int j = 0; j < n; ++j){
                if (grid[i][j] == 0){
                    continue;
                }
                int total = 0;
                vector<pair<int, int>> buff;
                buff.push_back({i, j});
                while (!buff.empty()){
                    auto [x, y] = buff.back();
                    buff.pop_back();
                    if (grid[x][y] == 0){
                        continue;
                    }
                    total += grid[x][y];
                    grid[x][y] = 0;
                    for (int d = 0; d < 4; ++d){
                        int xx = x + this->DIRS[d][0],
                            yy = y + this->DIRS[d][1];
                        if (0 <= xx && xx < m && 0 <= yy && yy < n && grid[xx][yy] != 0){
                            buff.push_back({xx, yy});
                        }
                    }
                }
                res = max(res, total);
            }
        }
        return res;
    }
};
